rfcs/25-ui-systems-callbacks.md

```diff
@@ -0,0 +1,493 @@
+
+### Why do we want callbacks?
+
+UI, much like scripting, tends to involve a large number of special-cased behaviors, in direct opposition to the natural patterns promoted by the ECS architecture.
+The callback pattern (and more generally, hooks) allows us to express this logic in a sane and maintainable fashion.
+
+Theoretically, everything that users could do with the callback pattern could be done with one-off systems.
+However, this clutters our scheduler (possibly hurting performance), reduces clarity and hurts compile times due to a huge number of one-off types.
+Furthermore, by storing specialized logic as data directly on UI entities it makes it dramatically easier to debug and reason about customized behavior.
+
+### Why don't we need a more complex reactivity model?
+
+Most reaction chains in UI are shockingly short.
+The most complex patterns are:
+
+1. Changing UI appearance en-masse: styles and themes should be used for this.
+2. Changing which UI elements are displayed (e.g. swapping tabs, pulling up a menu): this is precisely what `on_enter` and `on_exit` system sets in `States` are intended to solve.
+3. Handling layout changes: this should be done automatically in a single system (or group of systems) that runs in `CoreStage::PostUpdate`, rather than being scatter across our logic.
+
+For everything else, change detection and events should be more than adequate.
+
+By sticking to a straightforward imperative model of UI behavior combined with decoupled reactivity,
+we can make complex user interfaces dramatically easier to reason about, refactor and debug.
+
+## Unresolved questions
+
+- How do we serialize and deserialize callback components?
+- How can we ergonomically control the order in which callback commands are executed?
+- Should we special-case UI callbacks for now, or create a generalized `Hook` trait from the very beginning?
+
+## Future work
+
+1. We may want to loop over UI in some way to ensure that everything is resolved properly.
+2. The `EntityCommand` variant of `Callback` may be better handled using `Relations` in some form in the future.
+3. The ergonomics and performance of the callback pattern will be improved with other possible improvements to commands, namely more immediate processing, parallel execution and better control over execution order.
+4. Accessing and modifying the behavior of other entities within a UI hierarchy *can* be done as is, but will be much more ergonomic with advanced relations features like the ability to query for data on the target entity.
+5. Create a `QueryCommands` type (which applies the command to every entity matching a given query), and then extend callbacks to support this use as well.
+
+### A generalized hook framework
+
+The callback pattern, once established for UI (or immediately, if there's appetite for it), can be easily extended to create powerful and expressive **entity hooks**.
```

Comment:
IMO, this should be integrated immediately, and describe UI as a special case.

@DavidVonDerau thinks "hooks" should be called "event handlers", which I agree is a simpler and more consistent name.

I'm going to try and split out the events and callbacks fields of this into two components for clarity / simplicity, allowing us to just use plain `Events<T>` components on our entity. I think that an `EventHandler<T>` wrapper that is added as a component (and then the corresponding systems are autoregistered in `.add_event`) is the way to go.